# CS 583-C-Homework 4
## Large language models

---

### ***Fill your details below***
### Name: Corey Heckel
### CWID: 10462028
### Email ID: checkel
### References: ***Cite your references here***


---
### Submission guidelines:

#### 1. Submit this notebook along with its PDF version. You can do this by clicking File->Print->"Save as PDF"

#### 2. Name the file as "<mailID_HWnumber.extension>".  

For example, mailID is abcdefg @stevens.edu then name the files as abcdefg_HW1.ipynb and abcdefg_HW1.pdf.

#### 3. Please do not Zip your files.

---

In [1]:
!pip install sentencepiece transformers accelerate einops


& was unexpected at this time.


In [3]:
import os
os.environ["KMP_DUPLICATE_LIB_OK"]="TRUE"
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
os.environ["KMP_DUPLICATE_LIB_OK"] = "TRUE"

torch.set_default_device("cpu")

model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct", torch_dtype="auto", trust_remote_code=True)
tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct", trust_remote_code=True)

question = "Stevens won $123.5 in a lottery. He paid 20% for the tax and he paid $5.2 for the processing fee. How much was he able to take home?"
prompt = "Please give me the answer to the following problem by subtracting the tax and the processing fee from the amount won: "

inputs = tokenizer(prompt+question, return_tensors="pt", return_attention_mask=False)

outputs = model.generate(**inputs, max_length=300)
text = tokenizer.batch_decode(outputs)[0]
print(text)


Please give me the answer to the following problem by subtracting the tax and the processing fee from the amount won: Stevens won $123.5 in a lottery. He paid 20% for the tax and he paid $5.2 for the processing fee. How much was he able to take home? To determine how much Stevens was able to take home after paying taxes and processing fees, we need to follow these steps:

1. Calculate the total amount of tax Stevens has to pay.
2. Subtract the tax from the amount won by Stevens.
3. Subtract the processing fee from the remaining amount.

First, let's calculate the tax. The tax rate is 20%, so we can find the tax amount as follows:
\[ \text{Tax} = 20\% \times \$123.5 = 0.20 \times 123.5 = \$24.70 \]

Next, we subtract the tax from the amount won by Stevens:
\[ \text{Amount left after tax} = \$123.5 - \$24.70 = \$98.80 \]

Now, we subtract the processing fee from the remaining amount:
\[ \text{Amount taken home} = \$98.80 - \$5.20 = \$93.60 \]

So, the amount that Stevens was able to take

### Q1 (60 points.)
#### Finish running the query and prompt, record and answer.
To determine how much Stevens was able to take home, we need to follow these steps:

1. Calculate the amount of tax Stevens has to pay.
2. Add the tax amount to the original prize money.

First, let's calculate the tax. The tax is 20% of$the $123.5 prize money. We can 20%d 20% of $123.5 by multiplying $123.5 by 0.20 (which is the decimal equivalent of 20%).

\[
\text{Tax} = 123.5 \times 0.20 = 24.7
\]

Next, we add the tax to the original prize money to find out how much Stevens takes home.

\[
\text{Total Amount} = 123.5 + 24.7 = 148.2
\]

So, the amount Stevens was able to take home is \(\boxed{148.2}\).<|endoftext|>
#### Answer the question: does the LLM ans
No the LLM does not answer it correctly, as the 20% tax would need to be subtracted from the prize money and the processing fee would also need to be subtracted.wer it correctly?

### Q2 (10 points.)
#### Revise the prompt message "Please give me the answer to the following problem: " to make the LLM answer the question correctly.
To determine how much Stevens was able to take home after paying taxes and processing fees, we need to follow these steps:

1. Calculate the total amount of tax Stevens has to pay.
2. Subtract the tax from the amount won by Stevens.
3. Subtract the processing fee from the remaining amount.

First, let's calculate the tax. The tax rate is 20%, so we can find the tax amount as follows:
\[ \text{Tax} = 20\% \times \$123.5 = 0.20 \times 123.5 = \$24.70 \]

Next, we subtract the tax from the amount won by Stevens:
\[ \text{Amount left after tax} = \$123.5 - \$24.70 = \$98.80 \]

Now, we subtract the processing fee from the remaining amount:
\[ \text{Amount taken home} = \$98.80 - \$5.20 = \$93.60 \]

So, the amount that Stevens was able to take home i

The LLM answers the question correctly with this prompt: Please give me the answer to the following problem by subtracting the tax and the processing fee from the amount won:s:
\[ \boxed{93

### Q3 (30 points.)
#### Please discuss ways to improve the LLM's capacity in answering this question.
You could improve the LLM's capacity in answering this question by providing more clear instructions especially regarding mathimatical equations. Another idea may be to fine tune the model on datasets that include word problems to better understand the steps required to accurately answer these types of questions. We need to limit the ambiguity relating to how to handle certain aspects of a word problem. The LLM made assumptions on what to do with the tax and added it to the win amount instead of subtracting it.

### References:
[1] Wei, Jason, Xuezhi Wang, Dale Schuurmans, Maarten Bosma, Fei Xia, Ed Chi, Quoc V. Le, and Denny Zhou. "Chain-of-thought prompting elicits reasoning in large language models." Advances in neural information processing systems 35 (2022): 24824-24837. (https://arxiv.org/abs/2201.11903).

[2] Chen, Jiuhai, Lichang Chen, Chen Zhu, and Tianyi Zhou. "How Many Demonstrations Do You Need for In-context Learning?." arXiv preprint arXiv:2303.08119 (2023). (https://arxiv.org/abs/2303.08119).
